### ISSUE TITLE
Reloading mechanism uses relative paths

### ISSUE THREAD

**Issue by tdavis** [adzerk-oss/boot-reload#74]
Though my first-load experience works fine, the reload mechanism attempts to load JS at relative URLs which fails. For instance:
1. I go to `localhost:3000/public/html/some-page.html`
2. I make a change to a CLJS file
3. `boot-reload` makes a request to `localhost:3000/public/html/public/js/something.js`

I tried all combinations of `asset-host` and `cljs-asset-path` in an attempt to resolve this, but mostly they caused more problems than they solved (for instance, `cljs-asset-path` without `asset-host` set makes requests to `http://public/js/something.js`).

Ultimately, I ended up making the following changes in `reload.cljs`:

``` clojure
(defn- absolute? [uri]
  (or (not (string/blank? (.getDomain uri))) (.startsWith (.getPath uri) "/")))

(defn- maybe-make-absolute [uri]
  (if-not (absolute? uri)
    (.setPath uri (str "/" (.getPath uri)))
    uri))

(defn- reload-js [changed {:keys [on-jsload]
                           :or {on-jsload identity}}]
  (let [js-files (filter #(ends-with? % ".js") changed)]
    (when (seq js-files)
      (-> #(-> % guri/parse .makeUnique maybe-make-absolute)
          ;; ...
      ))))
```

Offhand I'm not sure why relative path reloads would ever be desirable but I'm no expert on all the use cases of `boot-reload` so I don't know if this is a valid change for you to make. It worked well for me, though!

**Comment by Deraen** [comment]
Okay. Before deciding what is the real problem and solution here I need to understand your use-case.

Why is your page served as `localhost:3000/public/html/some-page.html`?
I presume when you publish your app this is not the URL path you want to present to users?
`public/html` part looks like something you use to organize the files, but that should not be presented to users. I think it is good idea to use similar URLs while developing the app as what you would use in real environment.

Often boot-reload is used with a application that is loaded from index.html on root of the domain. Other configurations should be possible, though I have not tested this in a while.

**Comment by tdavis** [comment]
> Why is your page served as localhost:3000/public/html/some-page.html?

It's served there primarily because it happens to be the structure of my resources folder. I could change my ``boot-http`` usage to serve it directly rather than traverse the path in my browser, which would make the URLs more similar to those in production. However, that wouldn't alleviate the issue for traditional websites that aren't SPAs. Consider the layout:

    -- index.html
    +- some_folder
       - foo.html

If I am at ``/some_folder/foo.html`` in my browser, the reload request would still go to ``/some_folder/js/foo.js`` rather than the intended ``/js/foo.js``. Does that help?

**Comment by tdavis** [comment]
As I continue to try to integrate boot-reload with my project, the more fundamental issue appears to be the assumption that resource paths are equivalent to asset paths (or can be made so by a prefix). This assumption falls apart pretty quickly in non-trivial environments.

My project contains multiple apps, each with their own static assets, ring handlers, etc. If they use a common scss framework, I don't want to copy that framework around my project. Instead, I just write the scss in ``resources/scss/`` and sift the output to the correct place during build, say ``<target>/public/sites/foo/css/main.css``. The ``foo`` app is then served on a particular port and all its static assets are available at ``/static/``.

This schema works fine until boot-reload comes and tries to reload ``/public/sites/foo/css/main.css`` which doesn't exist. I'm not sure where best to fix this: I can add sift-like semantics as a boot-reload argument instead of relying only on the simple ``:cljs-asset-path`` or I can add redirect routes in my application specifically to handle the issue in boot-reload. The latter would be easier, but it would create a coupling that really shouldn't exist.

Any thoughts on this?

**Comment by tdavis** [comment]
This bit me again today on a new project. Took me a couple hours to realize I needed to apply my patch. Here's a rundown of the project; it's much simpler this time. Here's my ``run``:

    (deftask run []
      (comp (serve)
        (watch)
        (reload :on-jsload 'dashboard.core/mount-components)
        (cljs-repl)
        (cljs :source-map true :optimizations :none)))

The dashboard app lives at ``src/cljs/dashboard`` with the following ``main.cljs.edn``:

    {:require  [dashboard.core]
     :init-fns [dashboard.core/init!]
     :compiler-options {}}

The index HTML simply includes ``<script src="main.js" type="text/javascript"></script>`` and is accessible from ``http://localhost:3000/dashboard/``.

On a change to the CLJS source, reload makes a request to ``http://localhost:3000/dashboard/dashboard/main.out/dashboard/core.js?zx=8bbc9b3tcb4u`` which 404s. Given that I've made no configuration changes, I'm struggling to see how boot-reload could _ever_ work from a non-root path, no matter how realistic it may map to a production URL.

**Comment by Deraen** [comment]
So this project has several apps with separate html files? And thus you can't set asset-path etc.?

**Comment by tdavis** [comment]
Correct. I could _technically_ set asset-path, though it would make development inconvenient and require separate servers for each app.

The problem is that even with asset-path set it's still going to reload requests relative to whatever page I'm on. So, say I set options such that I end up with ``/dashboard.js`` and ``/dashboard.out`` and change the script ``src`` to ``/dashboard.js``; reload will still make requests to ``http://localhost:3000/dashboard/dashboard.out/core.cljs`` which is still a 404. As far as I can tell, there's no set of options that allow reload to work at a non-root path; what the ``target`` directory looks like is ultimately irrelevant. Even cljs-asset-path is just tacked onto the beginning of the path, so setting it to "/" just results in requests to ``http://localhost:3000/dashboard//dashboard.out/core.cljs`` (double //)

**Comment by velveteer** [comment]
@tdavis Have you tried setting a base href in your pages?

**Comment by Deraen** [comment]
@tdavis You could try master branch, I think #83 might have changed this.

**Comment by Deraen** [comment]
And I deployed 0.4.10 with the change.

**Comment by Deraen** [comment]
I'll close this as inactive and presumable fixed by #83 in 0.4.10. If the problem still exists, please re-open with details.
